### PR TITLE
Restructure Garbage Collector

### PR DESCRIPTION
**Pull request by lushoBarlett**
Separates `GC::run` into different functions. Also implements `Environment::run_gc`

### CODE REVIEW

File: src/interpreter/environment.cpp [hunk 1]
```diff
@@ -56,42 +58,77 @@ Value* Environment::fetch_return_value() {
...
+{ return m_gc->new_dictionary(std::move(declarations)); }
+
+Function* Environment::new_function(FunctionType def, ObjectType s)
+{ return m_gc->new_function(def, std::move(s)); }
+
+NativeFunction* Environment::new_native_function(NativeFunctionType* fptr)
+{ return m_gc->new_native_function(fptr); }
+
+Error* Environment::new_error(std::string e)
+{ return m_gc->new_error(e); }
```
[the remark this answers is not in the document]
Comment (SebastianMestre) [reply src/interpreter/environment.cpp:125]: (I mean the new formatting)

File: src/interpreter/execute.cpp [hunk 2]
```diff
@@ -45,7 +45,7 @@ exit_status_type execute(std::string const& source, bool dump_ast, Runner* runne
 
 	exit_status_type runner_exit_code = runner(env);
 
-	gc.run();
```
[the remark this answers is not in the document]
Comment (SebastianMestre) [reply src/interpreter/execute.cpp:48]: Oh damn, I didn't actually grab the `gc.sweep_all()` line at all!

File: src/interpreter/environment.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (SebastianMestre) [reply src/interpreter/environment.cpp:125]: I'm not opposed to it, it just seems unnecessary

File: src/interpreter/environment.cpp [hunk 3]
```diff
@@ -56,42 +58,77 @@ Value* Environment::fetch_return_value() {
...
 
-// scope
+
+void Environment::run_gc() {
+	for (Scope* scope_it = m_scope; scope_it != nullptr; scope_it = scope_it->m_prev)
+		for (auto& p : scope_it->m_declarations)
+			gc_visit(p.second);
+
+	m_gc->sweep();
+}
```
[the remark this answers is not in the document]
Comment (SebastianMestre) [reply src/interpreter/environment.cpp:69]: Also, nice `taller de programación` skills

File: src/interpreter/environment.hpp [hunk 4]
```diff
@@ -34,27 +34,28 @@ struct Environment {
...
+	String*         new_string(std::string);
+	Array*          new_list(ArrayType);
+	Object*         new_object(ObjectType);
+	Dictionary*     new_dictionary(ObjectType);
+	Function*       new_function(FunctionType, ObjectType);
 	NativeFunction* new_native_function(NativeFunctionType*);
-	Error* new_error(std::string);
-	Reference* new_reference(Value*);
+	Error*          new_error(std::string);
+	Reference*      new_reference(Value*);
```
[the remark this answers is not in the document]
Comment (lushoBarlett) [reply src/interpreter/environment.hpp:58]: Ah yes, that is much better

File: src/interpreter/environment.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (lushoBarlett) [reply src/interpreter/environment.cpp:125]: Same thing with misaligned names, this was hard to read. So today I checked out a bit of STL code in the form of `make_unique` and saw this style and I decided that is was okay to use here (those were really long lines).

File: src/interpreter/environment.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (lushoBarlett) [reply src/interpreter/environment.cpp:69]: > I think we should call m_gc->unmark_all, followed by m_gc->mark_roots before iterating over the scopes here

You are absolutely right, I keep forgetting that gc will delete everything that is not visited.

> Also, nice taller de programación skills

Hahaha which part specifically made you think that?

File: src/interpreter/garbage_collector.cpp [hunk 5]
```diff
@@ -36,6 +40,11 @@ void GC::run() {
 	    std::remove_if(m_blocks.begin(), m_blocks.end(), is_null), m_blocks.end());
 }
 
+void GC::sweep_all() {
+	mark_roots();
+	sweep();
+}
```
[the remark this answers is not in the document]
Comment (lushoBarlett) [reply src/interpreter/garbage_collector.cpp:46]: Indeed. Continuing my previous comment, I also have the operations backwards is my head.

File: src/interpreter/garbage_collector.hpp [hunk 6]
```diff
@@ -18,21 +18,25 @@ struct GC {
...
+	Dictionary*     new_dictionary(ObjectType);
+	Array*          new_list(ArrayType);
+	Integer*        new_integer(int);
+	Float*          new_float(float);
+	Boolean*        new_boolean(bool);
+	String*         new_string(std::string);
+	Function*       new_function(FunctionType, ObjectType);
+	NativeFunction* new_native_function(NativeFunctionType*);
+	Error*          new_error(std::string);
+	Reference*      new_reference(Value*);
```
[the remark this answers is not in the document]
Comment (lushoBarlett) [reply src/interpreter/garbage_collector.hpp:39]: Your suggestion seems better to me, I think the arrows are good for eye-parsing and separating the arguments from the return type

File: src/interpreter/execute.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (lushoBarlett) [reply src/interpreter/execute.cpp:48]: Seems like a fine idea to me. Adding `sweep_all` to the destructor is good in and of itself.

File: src/interpreter/environment.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (SebastianMestre) [reply src/interpreter/environment.cpp:69]: > Hahaha which part specifically made you think that?

Walking a linked list

File: src/interpreter/environment.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (SebastianMestre) [reply src/interpreter/environment.cpp:128]:
```c++
int main()
{
  __STL __code = __is_not_always__(_M_the_best, std::_InTermsOf{ __style});
}
```

[the remark this answers is not in the document]
Comment (lushoBarlett) [reply src/interpreter/environment.cpp:128]: I call straw man fallacy on that one. I could always have written a standard function body, but it seemed overkill for essentially short-hand arrow functions.

[the remark this answers is not in the document]
Comment (SebastianMestre) [reply src/interpreter/environment.cpp:128]: Just a bit of tongue-in-cheek

**Pull request merged**